### PR TITLE
Streams implementation

### PR DESCRIPTION
**Pull request by H-Plus-Time**
Considerably cleaned up, the api for this is wholly functional thanks to wasm-streams (and the stream! macro from async-stream):
```js
const stream = readFFIStream(targetUrl);
const batches = []
for await (const ffiTable of stream) {
    const batch = parseRecordBatch(memory.buffer, ffiTable.arrayAddr(), ffiTable.schemaAddr(), true);
    batches.push(batch);
}
```

`parseRecordBatch` now works for both arrow1 and arrow2, as far as I can tell given the ffi tests, it should be fully compatible.

Two things came up wrt the typescript interface:
1. The bindings to ReadableStream are opaque, so their TS equivalents are just `ReadableStream<R = any>`. There's a combination of wasm_bindgen arguments that override a function's return type by fiat.
2. The much bigger problem - apparently all versions of Typescript (yes, even nightlies) still interpret WHATWG ReadableStreams as both a. lacking a [Symbol.asyncIterator]() and b. being non-iterable. Even once I forced the types to solve for 1, I ended up having to do `as unknown as FFIArrowRecordBatch[]`.

NB: I've opted to leave the write half of this for a future PR (post-integration with the arrow-wasm repo), and of course IPC stream outputs aren't particularly relevant anymore.

### CODE REVIEW

**Comment by kylebarron** [comment]
Awesome, thanks! I'll do a final review and merge tonight or tomorrow!

File: src/arrow1/ffi.rs [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+
+    #[wasm_bindgen]
+    pub fn free(self) {
+        drop(self.0)
+    }
+
+    #[wasm_bindgen]
+    pub fn drop(self) {
+        drop(self.0)
+    }
```
Comment (kylebarron) [inline src/arrow1/ffi.rs:28]: General note: I need to check that the memory pointed to by this _actually_ gets freed by wasm-bindgen. I only need one method but I'm not sure if I can/need to add a manual `drop` call to release the array

File: src/arrow1/reader_async.rs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    let content_length = match content_length {
+        Some(_content_length) => _content_length,
+        None => get_content_length(url.clone()).await?,
+    };
+    let reader = crate::common::fetch::create_reader(url, content_length, None);
+
+    let builder = ParquetRecordBatchStreamBuilder::new(reader.compat()).await?;
+    let parquet_reader = builder.build()?;
+    Ok(parquet_reader)
+}
```
Comment (kylebarron) [inline src/arrow1/reader_async.rs:79]: This is so clean!

File: src/arrow2/reader_async.rs [hunk 3]
```diff
@@ -117,18 +92,50 @@ pub async fn read_row_group(
...
-
-    writer.finish()?;
     Ok(output_file)
 }
+
+pub async fn read_record_batch_stream(
+    url: String,
+) -> Result<impl futures::Stream<Item = super::ffi::FFIArrowRecordBatch>> {
+    use async_stream::stream;
+    let inner_stream = stream! {
```
Comment (kylebarron) [inline src/arrow2/reader_async.rs:123]: woah! That's cool! I didn't even know `yield` was a keyword

**Review by kylebarron — APPROVED** [review]

File: src/arrow2/reader_async.rs (hunk 3 above)
Comment (H-Plus-Time) [reply src/arrow2/reader_async.rs:123]: The cake (`yield`) is a lie with this one - that stream! macro slightly extends the grammar (apparently it's on it's way to the language - generator syntax seems to be one of the harder ones to add to a language).

File: scripts/build.sh [hunk 4]
```diff
@@ -19,7 +19,9 @@ wasm-pack build \
   --out-dir tmp_build/node \
   --out-name arrow1 \
   --target nodejs \
-  $FLAGS
+  --features async \
+  $FLAGS &
+[ -n "$CI" ] && wait;
```
[the remark this answers is not in the document]
Comment (H-Plus-Time) [reply scripts/build.sh:24]: Yeah, turning the parallel behaviour _off_ in CI took a bit of bash-fu (conditionally backgrounding a command isn't possible, but nobody said anything about conditionally running `wait` :wink: ).

Not much point for arrow2 (the wasm-opt step for arrow2 builds is quite short, while the compile step is highly parallelized, so trying to run that in parallel ended up slowing things down).

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -66,7 +66,7 @@ jobs:
 
       - uses: actions/setup-node@v2
         with:
-          node-version: "16"
+          node-version: "18"
```
[the remark this answers is not in the document]
Comment (H-Plus-Time) [reply .github/workflows/test.yml:69]: Yep, turns out the ReadableStream constructor only showed up in Node 18. There is indeed a poly/ponyfill, written by a familar face: [MattiasBuelens/web-streams-polyfill](https://github.com/MattiasBuelens/web-streams-polyfill).

That does bring up a good point - overall README level documentation, and doc-strings for the new exports. On that note, the codeblocks embedded in rust doc comments, do those show up with syntax highlighting for embedded code fragments (a bit like jsdocs) in editors for you?

[the remark this answers is not in the document]
Comment (kylebarron) [reply .github/workflows/test.yml:69]: Yeah they show up in editors and on the docs site https://kylebarron.dev/parquet-wasm/modules/bundler_arrow1.html#readParquet

File: scripts/build.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (kylebarron) [reply scripts/build.sh:24]: I wonder why wasm-opt would be so much slower for arrow1. I don't think this used to be the case; I only recently noticed it taking longer

File: src/arrow2/reader_async.rs (hunk 3 above)
Comment (kylebarron) [reply src/arrow2/reader_async.rs:123]: Ah I see it's a proc macro so it can make up new keywords!

**Comment by kylebarron** [comment]
Thanks again! As a follow up I'll try to update some docs and integrate with the new arrow-wasm Table object, but it would be great to release this sooner than later

**Pull request merged**